lightning/restore/restore.go

```diff
@@ -307,48 +308,180 @@ outside:
+						tblName:  viewMeta.Name,
+						stmtType: schemaCreateView,
+						sql:      sql,
+					})
 				}
-				err := InitSchema(ctx, rc.tidbGlue, dbMeta.Name, viewsSchema)
+			}
+		}
+		worker.wg.Add(1)
+		worker.jobCh <- restoreSchemaJob
+	}
+	worker.wg.Wait()
+	worker.errCh <- nil
+}
 
+func (worker *restoreSchemaWorker) doJob() {
+	for {
+		select {
+		case <-worker.ctx.Done():
+			return
+		case job := <-worker.jobCh:
+			if job == nil {
+				return
+			}
+			var logger log.Logger
+			purposePicker := 0
+			for _, stmt := range job.stmts {
+				if stmt.stmtType == schemaCreateDatabase {
+					logger = log.With(zap.String("db", job.dbName))
+					purposePicker = 1
+				} else if stmt.stmtType == schemaCreateTable {
+					logger = log.With(zap.String("table", common.UniqueTable(job.dbName, stmt.tblName)))
+					purposePicker = 2
+				} else if stmt.stmtType == schemaCreateView {
+					logger = log.With(zap.String("table", common.UniqueTable(job.dbName, stmt.tblName)))
+					purposePicker = 3
+				}
+				task := logger.Begin(zap.DebugLevel, fmt.Sprintf("execute SQL: %s", stmt.sql))
+				err := worker.executor.ExecuteWithLog(worker.ctx, stmt.sql, worker.purpose[purposePicker], logger)
 				task.End(zap.ErrorLevel, err)
 				if err != nil {
-					return errors.Annotatef(err, "restore view schema %s failed", dbMeta.Name)
+					switch stmt.stmtType {
+					case schemaCreateDatabase:
+						err = errors.Annotatef(err, "restore database schema %s failed", job.dbName)
+					case schemaCreateTable:
+						err = errors.Annotatef(err, "restore table schema %s failed", stmt.tblName)
+					case schemaCreateView:
+						err = errors.Annotatef(err, "restore view schema %s failed", stmt.tblName)
+					}
+					worker.errCh <- err
+					worker.wg.Done()
+					return
 				}
 			}
+			worker.wg.Done()
+		}
+	}
+}
+
+func (worker *restoreSchemaWorker) wait() error {
+	defer worker.quit()
+	select {
+	case err := <-worker.errCh:
+		return err
+	case <-worker.ctx.Done():
+		return nil
+	}
+}
 
+func (rc *RestoreController) restoreSchema(ctx context.Context) error {
+	if !rc.cfg.Mydumper.NoSchema {
+		if rc.tidbGlue.OwnsSQLExecutor() {
+			// CONFUSED: why don't use rc.tidbGlue.GetDB()?
+			db, err := DBFromConfig(rc.cfg.TiDB)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			defer db.Close()
+			db.ExecContext(ctx, "SET SQL_MODE = ?", rc.cfg.TiDB.StrSQLMode)
+		}
+		concurrency := 16
+		childCtx, cancel := context.WithCancel(ctx)
+		worker := restoreSchemaWorker{
+			ctx:      childCtx,
+			quit:     cancel,
+			wg:       sync.WaitGroup{},
+			jobCh:    make(chan *schemaJob, concurrency),
+			errCh:    make(chan error),
+			parser:   rc.tidbGlue.GetParser(),
+			executor: rc.tidbGlue.GetSQLExecutor(),
+			store:    rc.store,
+			purpose: []string{
+				"unknown",
+				"create database",
+				"create table",
+				"create view",
+			},
+		}
+		go worker.makeJobs(rc.dbMetas)
+		for i := 0; i < concurrency; i++ {
+			go worker.doJob()
```

Comment:
IMO we should not let different phase(DB/table/view) concurrently running, because CREATE VIEW may reference tables in another DB and that table may not created.

and this PR won't help a lot when importing a big DB that containing many tables.

Reply:
I could use `worker.wg.Wait() waiting all create db|table statements executed.